File: ios/VoiceChatbot/VoiceChatAPI.swift
import CryptoKit
import Foundation
import Security

enum APIError: LocalizedError {
    case invalidServerURL
    case invalidResponse
    case server(Int, String)
    case certificateRequired
    case certificateMismatch

    var errorDescription: String? {
        switch self {
        case .invalidServerURL: "The server URL is invalid."
        case .invalidResponse: "The server returned an invalid response."
        case .server(let status, _): "Server error \(status). The PC could not complete the request."
        case .certificateRequired: "Import the PC server certificate before connecting."
        case .certificateMismatch: "The server certificate does not match the imported certificate."
        }
    }
}

final class PinnedCertificateDelegate: NSObject, URLSessionDelegate, @unchecked Sendable {
    private let pinnedDigest: Data?

    init(certificateDER: Data?) {
        pinnedDigest = certificateDER.map { Data(SHA256.hash(data: $0)) }
    }

    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let trust = challenge.protectionSpace.serverTrust,
              let certificate = SecTrustCopyCertificateChain(trust) as? [SecCertificate],
              let leaf = certificate.first
        else {
            completionHandler(.performDefaultHandling, nil)
            return
        }

        guard let pinnedDigest else {
            completionHandler(.performDefaultHandling, nil)
            return
        }

        let serverDER = SecCertificateCopyData(leaf) as Data
        let serverDigest = Data(SHA256.hash(data: serverDER))
        guard serverDigest == pinnedDigest else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }

        // The user explicitly imported and pinned this exact server certificate.
        // Make that certificate the trust anchor and evaluate it as plain X.509,
        // without tying trust to either the LAN or Tailscale IP address.
        let policyStatus = SecTrustSetPolicies(trust, SecPolicyCreateBasicX509())
        let anchorStatus = SecTrustSetAnchorCertificates(trust, [leaf] as CFArray)
        let anchorOnlyStatus = SecTrustSetAnchorCertificatesOnly(trust, true)
        guard policyStatus == errSecSuccess,
              anchorStatus == errSecSuccess,
              anchorOnlyStatus == errSecSuccess,
              SecTrustEvaluateWithError(trust, nil)
        else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }

        completionHandler(.useCredential, URLCredential(trust: trust))
    }
}

actor VoiceChatAPI {
    private let profile: ServerProfile
    private let baseURL: String
    private let session: URLSession

    init(profile: ServerProfile, baseURL: String, probeMode: Bool = false) {
        self.profile = profile
        self.baseURL = baseURL
        let configuration = probeMode
            ? URLSessionConfiguration.ephemeral
            : URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = probeMode ? 8 : 90
        configuration.timeoutIntervalForResource = probeMode ? 10 : 300
        configuration.waitsForConnectivity = true
        configuration.allowsCellularAccess = true
        configuration.allowsExpensiveNetworkAccess = true
        configuration.allowsConstrainedNetworkAccess = true
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        session = URLSession(
            configuration: configuration,
            delegate: PinnedCertificateDelegate(certificateDER: profile.pinnedCertificateDER),
            delegateQueue: nil
        )
    }

    func status() async throws -> ServerStatus {
        try await request(path: "/api/status")
    }

    func transcribe(wav: Data) async throws -> String {
        let result: TranscriptionResponse = try await uploadWav(wav, path: "/api/transcribe")
        return result.transcript.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func detectSpeech(wav: Data) async throws -> Bool {
        let result: SpeechDetectionResponse = try await uploadWav(wav, path: "/api/vad")
        return result.speech
    }

    private func uploadWav<T: Decodable>(_ wav: Data, path: String) async throws -> T {
        let boundary = UUID().uuidString
        var body = Data()
        body.append("--\(boundary)\r\n")
        body.append("Content-Disposition: form-data; name=\"audio\"; filename=\"iphone.wav\"\r\n")
        body.append("Content-Type: audio/wav\r\n\r\n")
        body.append(wav)
        body.append("\r\n--\(boundary)--\r\n")

        var request = try makeRequest(path: path, method: "POST")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.httpBody = body
        return try await perform(request)
    }

    func respond(to text: String, keepDocumentsActive: Bool = false) async throws -> AssistantResponse {
        var request = try makeRequest(path: "/api/respond", method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(
            TextRequest(text: text, keepDocumentsActive: keepDocumentsActive)
        )
        return try await perform(request)
    }

    func tool(prompt: String) async throws -> String {
        var request = try makeRequest(path: "/api/tool", method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(ToolRequest(prompt: prompt))
        let result: ToolResponse = try await perform(request)
        return result.response
    }

    func prepareTextMessage(_ text: String) async throws -> TextMessagePreparation {
        var request = try makeRequest(path: "/api/text-message", method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(TextMessagePreparationRequest(text: text))
        return try await perform(request)
    }

    func prepareCalendarEvent(
        text: String,
        currentDateTime: String,
        timeZone: String
    ) async throws -> CalendarAIDraft {
        var request = try makeRequest(path: "/api/calendar-draft", method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(
            CalendarPreparationRequest(
                text: text,
                currentDateTime: currentDateTime,
                timeZone: timeZone
            )
        )
        return try await perform(request)
    }

    func groundedAnswer(prompt: String) async throws -> String {
        var request = try makeRequest(path: "/api/grounded-answer", method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(ToolRequest(prompt: prompt))
        let result: GroundedAnswerResponse = try await perform(request)
        return result.answer
    }

    func krea2Options() async throws -> Krea2Options {
        try await request(path: "/api/krea2/options")
    }

    func createKrea2Image(
        prompt: String,
        enableLora: Bool,
        loraName: String,
        aspectRatio: String
    ) async throws -> AssistantResponse {
        var request = try makeRequest(path: "/api/krea2/create", method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(
            Krea2CreateRequest(
                prompt: prompt,
                enableLora: enableLora,
                loraName: loraName,
                aspectRatio: aspectRatio
            )
        )
        return try await perform(request)
    }

    func sendMessage(
        text: String,
        attachments: [PendingAttachment],
        keepDocumentsActive: Bool
    ) async throws -> AssistantResponse {
        let boundary = "VoiceChatbot-\(UUID().uuidString)"
        var body = Data()
        body.appendMultipartField(name: "text", value: text, boundary: boundary)
        body.appendMultipartField(
            name: "keepDocumentsActive",
            value: keepDocumentsActive ? "true" : "false",
            boundary: boundary
        )
        for attachment in attachments {
            body.appendMultipartFile(
                name: "files",
                filename: attachment.name,
                mimeType: attachment.mimeType,
                data: attachment.data,
                boundary: boundary
            )
        }
        body.append("--\(boundary)--\r\n")

        var request = try makeRequest(path: "/api/message", method: "POST")
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.httpBody = body
        return try await perform(request)
    }

    func speak(_ text: String) async throws -> String {
        var request = try makeRequest(path: "/api/speak", method: "POST")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(SpeakRequest(text: text))
        let result: SpeakResponse = try await perform(request)
        return result.audioURL
    }

    func audioData(relativePath: String) async throws -> Data {
        try await mediaData(relativePath: relativePath)
    }

    func mediaData(relativePath: String) async throws -> Data {
        let request = try makeRequest(path: relativePath, method: "GET")
        let (data, response) = try await session.data(for: request)
        try validate(response: response, data: data)
        return data
    }

    private func request<T: Decodable>(path: String) async throws -> T {
        try await perform(makeRequest(path: path, method: "GET"))
    }

    private func perform<T: Decodable>(_ request: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: request)
        try validate(response: response, data: data)
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func makeRequest(path: String, method: String) throws -> URLRequest {
        guard var base = URL(string: baseURL) else {
            throw APIError.invalidServerURL
        }
        if !path.isEmpty {
            base.append(path: path.trimmingCharacters(in: CharacterSet(charactersIn: "/")))
        }
        var request = URLRequest(url: base)
        request.httpMethod = method
        if !profile.pin.isEmpty {
            request.setValue(profile.pin, forHTTPHeaderField: "X-Phone-Remote-Pin")
        }
        return request
    }

    private func validate(response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }
        guard (200..<300).contains(http.statusCode) else {
            throw APIError.server(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
    }
}

private struct TextRequest: Encodable {
    let text: String
    let keepDocumentsActive: Bool
}

private struct ToolRequest: Encodable {
    let prompt: String
}

private struct ToolResponse: Decodable {
    let response: String
}

private struct TextMessagePreparationRequest: Encodable {
    let text: String
}

private struct CalendarPreparationRequest: Encodable {
    let text: String
    let currentDateTime: String
    let timeZone: String
}

private struct GroundedAnswerResponse: Decodable {
    let answer: String
}

struct Krea2Options: Decodable, Equatable {
    let loras: [String]
    let aspectRatios: [String]

    enum CodingKeys: String, CodingKey {
        case loras
        case aspectRatios
        case pascalLoras = "Loras"
        case pascalAspectRatios = "AspectRatios"
    }

    init(loras: [String], aspectRatios: [String]) {
        self.loras = loras
        self.aspectRatios = aspectRatios
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        loras = try container.decodeIfPresent([String].self, forKey: .loras)
            ?? container.decodeIfPresent([String].self, forKey: .pascalLoras)
            ?? []
        aspectRatios = try container.decodeIfPresent([String].self, forKey: .aspectRatios)
            ?? container.decodeIfPresent([String].self, forKey: .pascalAspectRatios)
            ?? Krea2Options.fallback.aspectRatios
    }

    static let fallback = Krea2Options(
        loras: [],
        aspectRatios: [
            "1:1 (Square)",
            "3:2 (Photo)",
            "4:3 (Standard)",
            "16:9 (Widescreen)",
            "21:9 (Ultrawide)",
            "2:3 (Portrait Photo)",
            "3:4 (Portrait Standard)",
            "9:16 (Portrait Widescreen)"
        ]
    )
}

private struct Krea2CreateRequest: Encodable {
    let prompt: String
    let enableLora: Bool
    let loraName: String
    let aspectRatio: String
}

struct TextMessagePreparation: Decodable {
    let isTextMessage: Bool
    let recipient: String
    let body: String
    let needsClarification: Bool
    let clarificationQuestion: String
}

private struct SpeechDetectionResponse: Decodable {
    let speech: Bool
}

private struct SpeakRequest: Encodable {
    let text: String
}

private struct SpeakResponse: Decodable {
    let audioURL: String

    enum CodingKeys: String, CodingKey {
        case audioURL = "audioUrl"
    }
}

private extension Data {
    mutating func append(_ string: String) {
        append(Data(string.utf8))
    }

    mutating func appendMultipartField(name: String, value: String, boundary: String) {
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n")
        append(value)
        append("\r\n")
    }

    mutating func appendMultipartFile(
        name: String,
        filename: String,
        mimeType: String,
        data: Data,
        boundary: String
    ) {
        let safeFilename = filename
            .replacingOccurrences(of: "\"", with: "")
            .replacingOccurrences(of: "\r", with: "")
            .replacingOccurrences(of: "\n", with: "")
        append("--\(boundary)\r\n")
        append("Content-Disposition: form-data; name=\"\(name)\"; filename=\"\(safeFilename)\"\r\n")
        append("Content-Type: \(mimeType)\r\n\r\n")
        append(data)
        append("\r\n")
    }
}

File: android/app/src/main/java/com/voicechatbot/android/VoiceChatApi.kt
package com.voicechatbot.android

import android.content.Context
import android.net.Uri
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.security.SecureRandom
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

class VoiceChatApi(
    private val context: Context,
    private val profile: ServerProfile,
    private val baseUrl: String
) {
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }
    private val client = unsafeClient()

    suspend fun status(): ServerStatus = get("/api/status")

    suspend fun transcribe(wav: File): String {
        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("audio", "android.wav", wav.asRequestBody("audio/wav".toMediaType()))
            .build()
        return postMultipart<TranscriptionResponse>("/api/transcribe", body).transcript.trim()
    }

    suspend fun detectSpeech(wav: File): Boolean {
        val body = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("audio", "android.wav", wav.asRequestBody("audio/wav".toMediaType()))
            .build()
        return postMultipart<SpeechDetectionResponse>("/api/vad", body).speech
    }

    suspend fun respond(text: String, keepDocumentsActive: Boolean): AssistantResponse {
        return postJson(
            "/api/respond",
            mapOf("text" to text, "keepDocumentsActive" to keepDocumentsActive)
        )
    }

    suspend fun message(
        text: String,
        attachments: List<PendingAttachment>,
        keepDocumentsActive: Boolean
    ): AssistantResponse = withContext(Dispatchers.IO) {
        val builder = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("text", text)
            .addFormDataPart("keepDocumentsActive", keepDocumentsActive.toString())

        attachments.forEach { attachment ->
            val file = copyUriToCache(attachment.uri, attachment.name)
            builder.addFormDataPart(
                "files",
                attachment.name,
                file.asRequestBody(attachment.mimeType.toMediaType())
            )
        }
        postMultipart("/api/message", builder.build())
    }

    suspend fun tool(prompt: String): String {
        val result: ToolResponse = postJson("/api/tool", mapOf("prompt" to prompt))
        return result.response
    }

    suspend fun groundedAnswer(prompt: String): String {
        val result: GroundedAnswerResponse = postJson("/api/grounded-answer", mapOf("prompt" to prompt))
        return result.answer
    }

    suspend fun speak(text: String): String {
        val result: SpeakResponse = postJson("/api/speak", mapOf("text" to text))
        return result.audioUrl
    }

    suspend fun prepareTextMessage(text: String): TextMessagePreparation {
        return postJson("/api/text-message", mapOf("text" to text))
    }

    suspend fun prepareCalendarEvent(
        text: String,
        currentDateTime: String,
        timeZone: String
    ): CalendarAIDraft {
        return postJson(
            "/api/calendar-draft",
            mapOf("text" to text, "currentDateTime" to currentDateTime, "timeZone" to timeZone)
        )
    }

    suspend fun krea2Options(): Krea2Options {
        val obj: JsonObject = get("/api/krea2/options")
        fun strings(name: String, fallback: String): List<String> {
            val element = obj[name] ?: obj[fallback] ?: return emptyList()
            return runCatching { element.jsonArray.map { it.jsonPrimitive.content } }.getOrDefault(emptyList())
        }
        return Krea2Options(
            loras = strings("loras", "Loras"),
            aspectRatios = strings("aspectRatios", "AspectRatios").ifEmpty { Krea2Options.fallbackAspectRatios }
        )
    }

    suspend fun createKrea2Image(
        prompt: String,
        enableLora: Boolean,
        loraName: String,
        aspectRatio: String
    ): AssistantResponse {
        return postJson(
            "/api/krea2/create",
            mapOf(
                "prompt" to prompt,
                "enableLora" to enableLora,
                "loraName" to loraName,
                "aspectRatio" to aspectRatio
            )
        )
    }

    suspend fun mediaFile(relativeUrl: String, filename: String): File = withContext(Dispatchers.IO) {
        val request = requestBuilder(relativeUrl).get().build()
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("HTTP ${response.code}: ${response.body?.string().orEmpty()}")
            val out = File(context.cacheDir, filename)
            response.body?.byteStream()?.use { input ->
                out.outputStream().use { output -> input.copyTo(output) }
            }
            out
        }
    }

    private suspend inline fun <reified T> get(path: String): T = withContext(Dispatchers.IO) {
        val request = requestBuilder(path).get().build()
        client.newCall(request).execute().use { response ->
            decode(response.code, response.body?.string().orEmpty())
        }
    }

    private suspend inline fun <reified T> postJson(path: String, payload: Map<String, Any>): T {
        val body = jsonObject(payload).toString().toRequestBody("application/json".toMediaType())
        return withContext(Dispatchers.IO) {
            val request = requestBuilder(path).post(body).build()
            client.newCall(request).execute().use { response ->
                decode(response.code, response.body?.string().orEmpty())
            }
        }
    }

    private fun jsonObject(payload: Map<String, Any>): JsonElement = buildJsonObject {
        payload.forEach { (key, value) ->
            put(
                key,
                when (value) {
                    is Boolean -> JsonPrimitive(value)
                    is Number -> JsonPrimitive(value)
                    else -> JsonPrimitive(value.toString())
                }
            )
        }
    }

    private suspend inline fun <reified T> postMultipart(path: String, body: MultipartBody): T =
        withContext(Dispatchers.IO) {
            val request = requestBuilder(path).post(body).build()
            client.newCall(request).execute().use { response ->
                decode(response.code, response.body?.string().orEmpty())
            }
        }

    private inline fun <reified T> decode(code: Int, text: String): T {
        if (code !in 200..299) error("HTTP $code: $text")
        return json.decodeFromString(text)
    }

    private fun requestBuilder(pathOrUrl: String): Request.Builder {
        val url = if (pathOrUrl.startsWith("http")) {
            pathOrUrl
        } else {
            baseUrl.trimEnd('/') + "/" + pathOrUrl.trimStart('/')
        }
        return Request.Builder().url(url).apply {
            if (profile.pin.isNotBlank()) header("X-Phone-Remote-Pin", profile.pin)
        }
    }

    private fun copyUriToCache(uri: Uri, name: String): File {
        val safe = name.replace(Regex("[^A-Za-z0-9._-]"), "_")
        val out = File(context.cacheDir, safe)
        context.contentResolver.openInputStream(uri).use { input ->
            requireNotNull(input) { "Could not open $name" }
            out.outputStream().use { output -> input.copyTo(output) }
        }
        return out
    }

    private fun unsafeClient(): OkHttpClient {
        val trustManager = object : X509TrustManager {
            override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) = Unit
            override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) = Unit
            override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
        }
        val sslContext = SSLContext.getInstance("TLS").apply {
            init(null, arrayOf(trustManager), SecureRandom())
        }
        return OkHttpClient.Builder()
            .sslSocketFactory(sslContext.socketFactory, trustManager)
            .hostnameVerifier { _, _ -> true }
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(5, TimeUnit.MINUTES)
            .writeTimeout(5, TimeUnit.MINUTES)
            .build()
    }

    @kotlinx.serialization.Serializable
    private data class ToolResponse(val response: String = "")

    @kotlinx.serialization.Serializable
    private data class GroundedAnswerResponse(val answer: String = "")
}
